oauth2_http/java/com/google/auth/oauth2/ComputeEngineCredentials.java

```diff
@@ -133,27 +147,55 @@ public AccessToken refreshAccessToken() throws IOException {
     return new AccessToken(accessToken, new Date(expiresAtMilliseconds));
   }
 
-  /**
-   * Return whether code is running on Google Compute Engine.
-   */
-  static boolean runningOnComputeEngine(HttpTransportFactory transportFactory) {
-    try {
-      GenericUrl tokenUrl = new GenericUrl(METADATA_SERVER_URL);
-      HttpRequest request =
-          transportFactory.create().createRequestFactory().buildGetRequest(tokenUrl);
-      HttpResponse response = request.execute();
-      // Internet providers can return a generic response to all requests, so it is necessary
-      // to check that metadata header is present also.
-      HttpHeaders headers = response.getHeaders();
-      if (OAuth2Utils.headersContainValue(headers, "Metadata-Flavor", "Google")) {
-        return true;
+  /** Return whether code is running on Google Compute Engine. */
+  static boolean runningOnComputeEngine(
+      HttpTransportFactory transportFactory, DefaultCredentialsProvider provider) {
+    // If the environment has requested that we do no GCE checks, return immediately.
+    if (Boolean.parseBoolean(provider.getEnv(DefaultCredentialsProvider.NO_GCE_CHECK_ENV_VAR))) {
+      return false;
+    }
+
+    GenericUrl tokenUrl = new GenericUrl(getMetadataServerUrl(provider));
+    for (int i = 1; i <= MAX_COMPUTE_PING_TRIES; ++i) {
+      try {
+        HttpRequest request =
+            transportFactory.create().createRequestFactory().buildGetRequest(tokenUrl);
+        request.setConnectTimeout(COMPUTE_PING_CONNECTION_TIMEOUT_MS);
+        HttpResponse response = request.execute();
+        try {
+          // Internet providers can return a generic response to all requests, so it is necessary
+          // to check that metadata header is present also.
+          HttpHeaders headers = response.getHeaders();
+          return OAuth2Utils.headersContainValue(headers, "Metadata-Flavor", "Google");
+        } finally {
+          response.disconnect();
+        }
+      } catch (SocketTimeoutException expected) {
+        // Ignore logging timeouts which is the expected failure mode in non GCE environments.
+      } catch (IOException e) {
+        LOGGER.log(
+            Level.WARNING, "Failed to detect whether we are running on Google Compute Engine.", e);
       }
-    } catch (IOException expected) {
-      // ignore
     }
     return false;
   }
 
+  public static String getMetadataServerUrl(DefaultCredentialsProvider provider) {
+    String metadataServerAddress = provider.getEnv(DefaultCredentialsProvider.GCE_METADATA_HOST_ENV_VAR);
+    if (metadataServerAddress != null) {
+      return "http://" + metadataServerAddress;
+    }
+    return getMetadataServerUrl();
+  }
+
+  public static String getMetadataServerUrl() {
+    return DEFAULT_METADATA_SERVER_URL;
+  }
+
+  public static String getTokenServerEncodedUrl() {
+    return getMetadataServerUrl() + "/computeMetadata/v1/instance/service-accounts/default/token";
```

Comment:
Shouldn't this also rely on `getMetadataServerUrl(DefaultCredentialsProvider)`?

This implementation has changed the original call order from `OAuth2Utils.java` slightly since `getMetadataServerUrl()` relied on calling `getMetadataServerUrl(DefaultCredentialsProvider provider)` there.